File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Adapters/ZLink/Spots/Handlers/PlayActorJoinGameHandler.cs
using Zlink.Framework.Contracts.Codecs.Json;
using TicTacToe.Shared.Contracts;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Spots;
using Systems.Zlink;
using TicTacToe.Server.Play.Adapters.ZLink.Actors;

namespace TicTacToe.Server.Play.Adapters.ZLink.Spots.Handlers;

internal sealed class PlayActorJoinGameHandler(ILogger<PlayActorJoinGameHandler> logger)
    : IZLinkEntrySpotActorRequestHandler<PlayEntrySpot, PlayActor, JoinGameReq, JoinGameRes>
{
    public async ValueTask<JoinGameRes> HandleAsync(
        PlayEntrySpot entrySpot,
        PlayActor actor,
        ZLinkSpotActorRequestContext context,
        JoinGameReq message,
        CancellationToken cancellationToken)
    {
        _ = entrySpot;
        _ = context;
        logger.LogInformation(
            "actor: JoinGameReq received. actor={ActorId}, roomId={RoomId}",
            actor.ActorId,
            message.RoomId);

        var spotRid = RoutingId.From(message.RoomId);
        var joined = await actor.Context.JoinSpot(
                spotRid,
                new TicTacToeGameJoinReq(message.RoomId, actor.ActorId).ToJson())
            .Async(cancellationToken);

        var joinReply = joined.Reply.FromJson<TicTacToeGameJoinRes>();
        var reply = new JoinGameRes(joinReply.State);
        logger.LogInformation(
            "actor -> client: JoinGameRes returned. actor={ActorId}, roomId={RoomId}, mark={Mark}",
            actor.ActorId,
            reply.State.RoomId,
            reply.State.XActorId == actor.ActorId ? TicTacToeMarks.X : TicTacToeMarks.O);
        return reply;
    }
}

File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Spots/Handlers/play-actor-join-game-handler.ts
import { Injectable } from '@nestjs/common';
import type {
  ZLinkActor,
  ZLinkEntrySpotActorRequestHandler,
  ZLinkSpotActorRequestContext
} from '@zlink-systems/framework';
import type { PlayEntrySpot } from '../play-entry-spot';
import type {
  JoinGameReq,
  JoinGameRes,
  TicTacToeActor
} from '../../../../../../Shared/Contracts/messages';

type PlayJoinActor = TicTacToeActor & ZLinkActor;

@Injectable()
class PlayActorJoinGameHandler
  implements ZLinkEntrySpotActorRequestHandler<PlayEntrySpot, PlayJoinActor, JoinGameReq, JoinGameRes> {
  async handle(
    entrySpot: PlayEntrySpot,
    actor: PlayJoinActor,
    context: ZLinkSpotActorRequestContext,
    request: JoinGameReq
  ): Promise<JoinGameRes> {
    void context;
    const actorRef = actor.context.actorRef;
    if (actorRef === undefined) {
      throw new Error(`Actor '${actor.actorId}' has no ActorRef.`);
    }
    return await entrySpot.join(actorRef, actor, request.roomId);
  }
}

export { PlayActorJoinGameHandler };
